crates/ruff_linter/src/rules/pylint/rules/access_member_before_definition.rs

```diff
@@ -0,0 +1,154 @@
+use ast::{call_path::collect_call_path, Stmt};
+use ruff_diagnostics::{Diagnostic, FixAvailability, Violation};
+use ruff_macros::violation;
+use ruff_python_semantic::ScopeKind;
+
+use ruff_text_size::Ranged;
+use rustc_hash::FxHashSet;
+use similar::DiffableStr;
+
+use crate::checkers::ast::Checker;
+use ruff_python_ast::{self as ast, Expr};
+
+/// Checks for accesses on uninitialized class members.
+///
+/// ## Example
+/// ```python
+/// class Unicorn:
+//  def __init__(self, x):
+//     if self.x > 9000:  # [access-member-before-definition]
+//         pass
+//     self.x = x
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class Unicorn:
+//  def __init__(self, x):
+//     self.x = x
+//     if self.x > 9000:
+//         pass
+/// ```
+#[violation]
+pub struct AccessMemberBeforeDefinition {
+    member: String,
+}
+
+impl Violation for AccessMemberBeforeDefinition {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::None;
+
+    fn message(&self) -> String {
+        let AccessMemberBeforeDefinition { member } = self;
+        format!("Accessed `{member}` before it was initialized",)
+    }
+
+    fn message_formats() -> &'static [&'static str] {
+        todo!()
+    }
+}
+
+pub(crate) fn access_member_before_definition(checker: &mut Checker, expr: &Expr) {
+    let Expr::Attribute(ast::ExprAttribute { value, attr, .. }) = expr else {
+        return;
+    };
+
+    if let Some(call_path) = collect_call_path(value) {
+        // Ignore `cls` accesses.
+        if matches!(call_path.as_slice(), ["cls"]) {
+            return;
+        }
+    }
+
+    // return if we are not in a class
+    if !checker.semantic().current_scope_ids().any(|scope_id| {
+        let scope = checker.semantic().scopes.get(scope_id).unwrap();
+        scope.kind.is_class()
+    }) {
+        return;
+    }
+
+    let init_function = checker.semantic().current_scope_ids().find_map(|scope_id| {
+        let scope = checker.semantic().scopes.get(scope_id).unwrap();
+        match &scope.kind {
+            ScopeKind::Function(function) => {
+                if function.name.as_str() == "__init__" {
+                    Some(function)
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    });
+
+    let function_body = match init_function {
+        Some(function) => &function.body,
+        None => return,
+    };
+
+    let mut assigned_members: FxHashSet<&str> = FxHashSet::default();
```

Comment:
I feel this code should not be here because we calculate attributes every time.

Reply:
I might suggest reframing the rule to run on each _scope_, and then check if the current scope is an `__init__ ` method. `unused_arguments` is one example of that pattern.

https://github.com/astral-sh/ruff/blob/da275b85725f501443fe73d5e2a5cf182a541881/crates/ruff_linter/src/rules/flake8_unused_arguments/rules/unused_arguments.rs#L316

Reply:
(Gonna hold off on reviewing all the logic below since some of it could change based on that refactor. But it would mean we analyze all assignments in one pass, rather than analyzing each attribute access.)

Reply:
I changed to rule to run on every scope. But there is a small challenge in traversing all the AST inside this function and finding out what attributes where assigned to self and were accessed. I'm thinking about moving that part to the checker so this data is populated like binding and here I just check for presence of accessed values.